[runtime] MonoError-ize mono_object_isinst{,_mbyref}

lambdageek opened the pull request:
Also mark them external only.  Runtime should use mono_object_isinst_mbyref_checked and mono_object_isinst_checked

kumpera commented on mono/metadata/icall.c:
```diff
@@ -250,6 +253,7 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
 
 #define INVALID_CAST G_STMT_START{ \
 		mono_get_runtime_callbacks ()->set_cast_details (vc, ec); \
+		mono_error_cleanup (&error);				\
```
We should raise the underlying exception and not hide it under an InvalidCastException

kumpera commented on mono/metadata/icall.c:
```diff
@@ -289,19 +293,23 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
 		return;
 	}
+	if (!is_ok (&error)) {
+		mono_error_cleanup (&error);
```
This is cleaning up the same error twice.

We should kill that INVALID_CAST macro and replace with something that takes no implicit state.

kumpera commented on mono/metadata/icall.c:
```diff
@@ -3063,7 +3076,8 @@ ves_icall_InternalInvoke (MonoReflectionMethod *method, MonoObject *this_arg, Mo
 		if (this_arg) {
-			if (!mono_object_isinst (this_arg, m->klass)) {
+			if (!mono_object_isinst_checked (this_arg, m->klass, &error)) {
+				mono_error_cleanup (&error);
```
We shouldn't swallow this error.

kumpera commented on mono/mini/exceptions-x86.c:
```diff
@@ -473,13 +474,14 @@ mono_x86_throw_exception (mgreg_t *regs, MonoObject *exc,
 			mono_ex->trace_ips = NULL;
 		}
 	}
+	mono_error_cleanup (&error);
```
This should be an assert, we shouldn't fail type loading for allocated objects.

kumpera commented on mono/metadata/object.c:
```diff
@@ -5970,18 +5970,45 @@ mono_object_unbox (MonoObject *obj)
 
+	MonoError error;
+	MonoObject *result = mono_object_isinst_checked (obj, klass, &error);
+	mono_error_cleanup (&error); /* FIXME better API that doesn't swallow the error */
```
If the method now is banned from the API, the comment can go.

kumpera commented on mono/mini/jit-icalls.c:
```diff
@@ -102,11 +102,13 @@ mono_ldvirtfn_gshared (MonoObject *obj, MonoMethod *method)
 	}
-	if (val && !mono_object_isinst (val, array->obj.vtable->klass->element_class)) {
+	if (val && !mono_object_isinst_checked (val, array->obj.vtable->klass->element_class, &error)) {
+		mono_error_cleanup (&error);
```
Don't swallow the error

kumpera commented on mono/mini/jit-icalls.c:
```diff
@@ -1192,8 +1195,9 @@ mono_object_castclass_unbox (MonoObject *obj, MonoClass *klass)
-	if (mono_object_isinst (obj, klass))
+	if (mono_object_isinst_checked (obj, klass, &error))
 		return obj;
+	mono_error_cleanup (&error);
```
Same thing, assert of throw, but don't swallow.

kumpera commented on mono/mini/jit-icalls.c:
```diff
@@ -1226,10 +1231,11 @@ mono_object_castclass_with_cache (MonoObject *obj, MonoClass *klass, gpointer *c
 		*cache = obj_vtable;
 		return obj;
 	}
+	mono_error_cleanup (&error);
```
Don't swallow the error.

kumpera commented on mono/mini/jit-icalls.c:
```diff
@@ -1257,10 +1264,11 @@ mono_object_isinst_with_cache (MonoObject *obj, MonoClass *klass, gpointer *cach
 		*cache = (gpointer)obj_vtable;
 		return obj;
 	} else {
+		mono_error_cleanup (&error);
```
We shouldn't swallow this error and silently fail the type check.

kumpera commented on mono/mini/mini-exceptions.c:
```diff
@@ -1373,10 +1374,11 @@ mono_handle_exception_internal_first_pass (MonoContext *ctx, MonoObject *obj, gi
 		mono_ex = (MonoException*)obj;
 		initial_trace_ips = mono_ex->trace_ips;
 	} else {
+		mono_error_cleanup (&error);
```
This should be an assert.

kumpera commented on mono/mini/mini-exceptions.c:
```diff
@@ -1605,7 +1608,8 @@ mono_handle_exception_internal (MonoContext *ctx, MonoObject *obj, gboolean resu
 
-	if (!mono_object_isinst (obj, mono_defaults.exception_class)) {
+	if (!mono_object_isinst_checked (obj, mono_defaults.exception_class, &error)) {
+		mono_error_cleanup (&error);
```
Make it an assert.

kumpera commented on mono/mini/mini-exceptions.c:
```diff
@@ -3051,10 +3059,13 @@ mono_llvm_match_exception (MonoJitInfo *jinfo, guint32 region_start, guint32 reg
 			break;
-		} else if (ei->flags == MONO_EXCEPTION_CLAUSE_FILTER) {
+		} else
+			mono_error_cleanup (&error);
```
Assert instead.

kumpera commented on mono/mini/mini-exceptions.c:
```diff
@@ -2882,7 +2889,8 @@ throw_exception (MonoObject *ex, gboolean rethrow)
 
-	if (!mono_object_isinst (ex, mono_defaults.exception_class)) {
+	if (!mono_object_isinst_checked (ex, mono_defaults.exception_class, &error)) {
+		mono_error_cleanup (&error);
```
Assert instead.

kumpera commented:
My understanding is that mono_object_isinst_checked and friends are functions that return 3 possible values: true, false and false + error.

We should not systematically ignore the last case.

lambdageek commented:
Thanks @kumpera.  I updated the PR to cover the false+error cases.

Also I missed all the non-OSX platforms, so I also updated the PR to fixup the other `exceptions-{arch}.c`  files.

lambdageek commented:
@monojenkins merge
